source/mqtt/Mqtt5Listener.cpp

```diff
@@ -0,0 +1,430 @@
+                /* release Mqtt5ListenerCore reference */
+                listener->m_selfReference.reset();
+            }
+
+            std::shared_ptr<Mqtt5Listener> Mqtt5Listener::NewMqtt5Listener(
+                const Mqtt5ListenerOptions &options,
+                const std::shared_ptr<Mqtt5Client> &client,
+                Allocator *allocator) noexcept
+            {
+                if (client == nullptr)
+                {
+                    AWS_LOGF_DEBUG(AWS_LS_MQTT5_GENERAL, "Failed to create Mqtt5Listener. Invalid Client.");
+                    aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+                    return nullptr;
+                }
+
+                /* Copied from MqttClient.cpp:ln754 (MqttClient::NewConnection) */
+                // As the constructor is private, make share would not work here. We do make_share manually.
+                Mqtt5Listener *toSeat =
+                    reinterpret_cast<Mqtt5Listener *>(aws_mem_acquire(allocator, sizeof(Mqtt5Listener)));
+                if (!toSeat)
+                {
+                    return nullptr;
+                }
+
+                toSeat = new (toSeat) Mqtt5Listener(options, client, allocator);
+                // Creation failed, make sure we release the allocated memory
+                if (!*toSeat)
+                {
+                    Crt::Delete(toSeat, allocator);
+                    return nullptr;
+                }
+
+                std::shared_ptr<Mqtt5Listener> shared_listener = std::shared_ptr<Mqtt5Listener>(
+                    toSeat, [allocator](Mqtt5Listener *listener) { Crt::Delete(listener, allocator); });
+                return shared_listener;
+            }
+
+            Mqtt5Listener::Mqtt5Listener(
+                const Mqtt5ListenerOptions &options,
+                const std::shared_ptr<Mqtt5Client> &client,
+                Allocator *allocator) noexcept
+                : m_allocator(allocator)
+            {
+                m_listener_core = Mqtt5ListenerCore::NewMqtt5ListenerCore(options, client, m_allocator);
+            }
+
+            void Mqtt5Listener::Subscribe(Crt::String topic, Crt::Mqtt5::OnPublishReceivedHandler callback) noexcept
+            {
+                if (m_listener_core != nullptr)
+                {
+                    m_listener_core->Subscribe(topic, callback);
+                }
+            }
+            void Mqtt5Listener::Unsubscribe(Crt::String topic) noexcept
+            {
+                if (m_listener_core != nullptr)
+                {
+                    m_listener_core->Unsubscribe(topic);
+                }
+            }
+            Mqtt5::Mqtt5Listener::~Mqtt5Listener()
+            {
+                if (m_listener_core != nullptr)
+                {
+                    m_listener_core->Close();
+                }
+            }
+
+            Mqtt5ListenerOptions::Mqtt5ListenerOptions() noexcept
+            {
+                onAttemptingConnect = NULL;
```

Comment:
Either default construct these explicitly in the initializer list or leave it out entirely.